pkg/operator/operator.go

```diff
@@ -241,76 +273,284 @@ func (k *Kontroller) withLeaderElection() error {
+			}
+		}
 	}
 
-	// Verify no nodes are still in the process of rebooting to avoid rebooting N > maxRebootingNodes
+	return nil
+}
+
+// checkAfterReboot gets all nodes with the after-reboot=true label and checks
+// if  all of the configured after-reboot annotations are set to true. If they
+// are, it deletes the after-reboot=true label and sets reboot-ok=false to tell
+// the agent that it has completed it's reboot successfully.
+// If there is an error getting the list of nodes or updating any of them, an
+// error is immediately returned.
+func (k *Kontroller) checkAfterReboot() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed listing nodes: %v", err)
+	}
+
+	postRebootNodes := k8sutil.FilterNodesByRequirement(nodelist.Items, afterRebootReq)
+
+	for _, n := range postRebootNodes {
+		if hasAllAnnotations(n, k.afterRebootAnnotations) {
+			glog.V(4).Infof("Deleting label %q for %q", constants.LabelAfterReboot, n.Name)
+			glog.V(4).Infof("Setting annotation %q to false for %q", constants.AnnotationOkToReboot, n.Name)
+			err = k8sutil.UpdateNodeRetry(k.nc, n.Name, func(node *v1api.Node) {
+				delete(node.Labels, constants.LabelAfterReboot)
+				// cleanup the after-reboot annotations
+				for _, annotation := range k.afterRebootAnnotations {
+					delete(node.Annotations, annotation)
+				}
+				node.Annotations[constants.AnnotationOkToReboot] = constants.False
+			})
+			if err != nil {
+				return fmt.Errorf("Failed to update node %q: %v", n.Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// markBeforeReboot gets nodes which want to reboot and marks them with the
+// before-reboot=true label. This is considered the beginning of the reboot
+// process from the perspective of the update-operator. It will only mark
+// nodes with this label up to the maximum number of concurrently rebootable
+// nodes as configured with the maxRebootingNodes constant.
+// It cleans up the before-reboot annotations before it applies the label, in
+// case there are any left over from the last reboot.
+// If there is an error getting the list of nodes or updating any of them, an
+// error is immediately returned.
+func (k *Kontroller) markBeforeReboot() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed listing nodes: %v", err)
+	}
+
+	// find nodes which are still rebooting
 	rebootingNodes := k8sutil.FilterNodesByAnnotation(nodelist.Items, stillRebootingSelector)
-	glog.V(6).Infof("Found %+v rebooting nodes", rebootingNodes)
+	// nodes running before and after reboot checks are still considered to be "rebooting" to us
+	beforeRebootNodes := k8sutil.FilterNodesByRequirement(nodelist.Items, beforeRebootReq)
+	rebootingNodes = append(rebootingNodes, beforeRebootNodes...)
+	afterRebootNodes := k8sutil.FilterNodesByRequirement(nodelist.Items, afterRebootReq)
+	rebootingNodes = append(rebootingNodes, afterRebootNodes...)
```

Comment:
I don't think this is a new issue, but we should indicate the lack of rebooting nodes if we're going to info log the count of rebooting nodes. It gives the appearance of being stuck at 1.